vms/shift/views.py

```diff
@@ -1,728 +1,832 @@
+        context['volunteer'] = get_volunteer_by_id(volunteer_id)
+        context['shift_list'] = get_unlogged_shifts_by_volunteer_id(
+            volunteer_id)
+        context['shift_list_with_hours'] = get_volunteer_shifts_with_hours(
+            volunteer_id)
+        return context
+
+
+@login_required
+def sign_up(request, shift_id, volunteer_id):
+    if shift_id:
+        shift = get_shift_by_id(shift_id)
+        if shift:
+
+            user = request.user
+            admin = None
+            volunteer = None
+
+            try:
+                admin = user.administrator
+            except ObjectDoesNotExist:
+                pass
+            try:
+                volunteer = user.volunteer
+            except ObjectDoesNotExist:
+                pass
+
+            if not admin and not volunteer:
+                return HttpResponse(status=403)
+
+            if volunteer:
+                if (int(volunteer.id) != int(volunteer_id)):
+                    return HttpResponse(status=403)
+
+            if request.method == 'POST':
+                try:
+                    result = register(volunteer_id, shift_id)
+                    if result == "IS_VALID":
+                        if admin:
+                            return HttpResponseRedirect(
+                                reverse(
+                                    'shift:manage_volunteer_shifts',
+                                    args=(volunteer_id, )))
+                        if volunteer:
+                            return HttpResponseRedirect(
+                                reverse(
+                                    'shift:view_volunteer_shifts',
+                                    args=(volunteer_id, )))
+                    else:
+                        return render(request, 'shift/sign_up_error.html', {
+                            'error_code': result
+                        })
+                except ObjectDoesNotExist:
+                    raise Http404
+            else:
+                return render(request, 'shift/sign_up.html', {
+                    'shift': shift,
+                    'volunteer_id': volunteer_id
+                })
+        else:
+            raise Http404
+    else:
+        raise Http404
+
+
+class ViewHoursView(LoginRequiredMixin, FormView, TemplateView):
+    template_name = 'shift/hours_list.html'
+    form_class = HoursForm
+
+    @method_decorator(check_correct_volunteer)
+    @method_decorator(vol_id_check)
+    def dispatch(self, *args, **kwargs):
+        return super(ViewHoursView, self).dispatch(*args, **kwargs)
+
+    def get_context_data(self, **kwargs):
+        context = super(ViewHoursView, self).get_context_data(**kwargs)
+        volunteer_id = self.kwargs['volunteer_id']
+        context['volunteer'] = get_volunteer_by_id(volunteer_id)
+        context['volunteer_shift_list'] = get_volunteer_shifts_with_hours(
+            volunteer_id)
+        context['last_date'] = timezone.now()

```

Comment:
Add the logic of 7 days here and give this key some more meaningful name like `init_date` as it will be the earliest date which will be considered for requesting edit.